source/common/jwt_authn/extractor.cc

```diff
@@ -0,0 +1,148 @@
+const std::string AuthorizationHeader = "authorization";
+
+// The authorization bearer prefix.
+const std::string BearerPrefix = "Bearer ";
+
+// The query parameter name to get JWT token.
+const std::string ParamAccessToken = "access_token";
+
+// A base JwtLocation object to store token and specified_issuers.
+class JwtLocationBase : public JwtLocation {
+public:
+  JwtLocationBase(const std::string& token, const std::set<std::string>& issuers)
+      : token_(token), specified_issuers_(issuers) {}
+
+  // Get the token string
+  const std::string& token() const override { return token_; }
+
+  // Check if an issuer has specified the location.
+  bool isIssuerSpecified(const std::string& issuer) const override {
+    return specified_issuers_.find(issuer) != specified_issuers_.end();
+  }
+
+private:
+  // Extracted token.
+  std::string token_;
+  // Stored issuers specified the location.
+  const std::set<std::string>& specified_issuers_;
+};
+
+// The JwtLocation for header extraction.
+class JwtHeaderLocation : public JwtLocationBase {
+public:
+  JwtHeaderLocation(const std::string& token, const std::set<std::string>& issuers,
+                    const Http::LowerCaseString& header)
+      : JwtLocationBase(token, issuers), header_(header) {}
+
+  void removeJwt(Http::HeaderMap& headers) const override { headers.remove(header_); }
+
+private:
+  // the header name the JWT is extracted from.
+  const Http::LowerCaseString& header_;
+};
+
+// The JwtLocation for param extraction.
+class JwtParamLocation : public JwtLocationBase {
+public:
+  JwtParamLocation(const std::string& token, const std::set<std::string>& issuers,
+                   const std::string& param)
+      : JwtLocationBase(token, issuers), param_(param) {}
+
+  void removeJwt(Http::HeaderMap&) const override {
+    // TODO(qiwzhang): remove JWT from parameter.
+  }
+
+private:
+  // the parameter key the JWT is extracted from.
+  const std::string& param_;
+};
+
+} // namespace
+
+Extractor::Extractor(const JwtAuthentication& config) {
+  for (const auto& rule : config.rules()) {
+    bool use_default = true;
+    if (rule.from_headers_size() > 0) {
+      use_default = false;
+      for (const auto& header : rule.from_headers()) {
+        config_header(rule.issuer(), header.name(), header.value_prefix());
+      }
+    }
+    if (rule.from_params_size() > 0) {
+      use_default = false;
+      for (const std::string& param : rule.from_params()) {
+        config_param(rule.issuer(), param);
+      }
+    }
+
+    // If not specified, use default locations.
+    if (use_default) {
+      config_header(rule.issuer(), AuthorizationHeader, BearerPrefix);
+      config_param(rule.issuer(), ParamAccessToken);
+    }
+  }
+}
+
+void Extractor::config_header(const std::string& issuer, const std::string& header_name,
+                              const std::string& value_prefix) {
+  auto& map_value = header_maps_[Http::LowerCaseString(header_name)];
```

Comment:
I think this is wrong, theoretically you might have same header name with different prefix.

Reply:
Updated to support same header name with different prefix.